File: 2023/12/part_one.php
#!/usr/bin/env php
<?php
$sum = 0;

$handle = fopen("input.txt", "r");
while (($line = fgets($handle)) !== false) {
	[$springs, $groups] = explode(" ", $line);
	$groups = array_map(
		"intval",
		explode(",", $groups)
	);
	test_line($springs, $groups);
}
fclose($handle);
echo $sum . "\n";

function test_line(string $springs, array $groups, int $run = 0): void {
	global $sum;
	for ($i = 0; $i < strlen($springs); $i++) {
		switch ($springs[$i]) {
			case ".":
				if ($run > 0) {
					if (isset($groups[0]) && $groups[0] == $run) {
						array_shift($groups);
						$run = 0;
					}
					else {
						return;
					}
				}
				break;
			case "#":
				if (!isset($groups[0])) {
					return;
				}
				$run++;
				break;
			case "?":
				test_line(
					"." . substr($springs, $i + 1),
					$groups,
					$run
				);
				test_line(
					"#" . substr($springs, $i + 1),
					$groups,
					$run
				);
				return;
				break;
			default:
				throw new Exception("Unknown spring type: " . $springs[$i]);
				break;
		}
		if (array_sum($groups) + count($groups) - 1 > strlen($springs) - $i + $run) {
			return;
		}
	}
	if (isset($groups[0]) && $groups[0] != $run) {
		return;
	}
	$sum++;
}

?>

File: 2023/15/part_one.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

int main() {
	int sum = 0;
	FILE* file = fopen("input.txt", "r");
	char c;
	int currentValue = 0;
	while ((c = fgetc(file)) != EOF) {
		if (c != '\n') {
			if (c != ',') {
				currentValue += c;
				currentValue *= 17;
				currentValue %= 256;
			}
			else {
				sum += currentValue;
				currentValue = 0;
			}
		}
	}
	fclose(file);
	sum += currentValue;

	printf("%d\n", sum);

	return 0;
}